### ISSUE TITLE
The consistency issue between before and after images

### ISSUE THREAD

**Issue by SensenGao** [multimodal-reasoning-lab/Bagel-Zebra-CoT#10]
I think the fine-tuned bagel model currently seems to have poor consistency and often makes mistakes in these simple image editing tasks. However, even with incorrect images, as long as the crossed text is correct, the model can still answer correctly. So I wonder whether the images are actually being used during reasoning.

Example 1:
<img width="1150" height="681" alt="Image" src="https://github.com/user-attachments/assets/a79a0ca2-78e2-4456-9f3c-c453c1761070" />

<img width="1760" height="1135" alt="Image" src="https://github.com/user-attachments/assets/9408a1a6-4474-44ee-b062-14d81788b457" />

<img width="1730" height="1106" alt="Image" src="https://github.com/user-attachments/assets/1b2b11d6-8326-4862-beb2-5862c1c18e77" />

Example 2:

<img width="1658" height="989" alt="Image" src="https://github.com/user-attachments/assets/f0bf6328-9b4a-4ed6-971a-3870aef93b11" />

<img width="1750" height="971" alt="Image" src="https://github.com/user-attachments/assets/c1b92551-599b-4b3a-917a-2196e408d425" />

<img width="1578" height="975" alt="Image" src="https://github.com/user-attachments/assets/36c34bde-8af4-4de9-940d-00e7aa37aaff" />

In both of these examples, the model gave the correct answer: 5.

**Comment by LeonLixyz** [comment]
Hi, thank you for trying our models and for providing your feedback!

Regarding the inconsistency: we believe that most of the image generation and editing capability comes from the pretraining stage, which was carried out (thanks to their remarkable work!) by the Bagel team. When we tested their model on the same task for image editing with thinking, we observed the following outputs:

<img width="768" height="512" alt="Image" src="https://github.com/user-attachments/assets/7adbe577-c25f-4947-a28c-3169ef720b4a" />
<img width="768" height="512" alt="Image" src="https://github.com/user-attachments/assets/f3bc5e61-f0fe-4b66-b563-2150f7e7dfe6" />
<img width="768" height="512" alt="Image" src="https://github.com/user-attachments/assets/8473c48b-0e56-40b0-9c79-d61d7a83b57c" />
<img width="768" height="512" alt="Image" src="https://github.com/user-attachments/assets/2ab60992-64a7-48c1-b7a5-f914ba419b89" />
<img width="768" height="512" alt="Image" src="https://github.com/user-attachments/assets/9bccc1f7-ea43-4695-88eb-3e4a0a24c650" />

In some sense, our model produces outputs with better factual accuracy when compared side by side.

Overall, our hypothesis is that the bulk of image generation ability originates from the pretraining stage, while our dataset is designed for post-training. Its goal is to encourage models to generate images as part of their own reasoning process, thereby enhancing visual reasoning.

Regrading whether the images are actually being used during reasoning: yes, we iteratively feed all generated images back into the context. We assume that those generated images helped with reasoning in the latent space, which is why though generated images are inconsistent, but the answers are the same. 

The faithfulness of CoT is an interesting open problem, checkout the following papers:
[Language Models Don't Always Say What They Think: Unfaithful Explanations in Chain-of-Thought Prompting](https://arxiv.org/abs/2305.04388)
[Reasoning Models Don't Always Say What They Think
](https://www.anthropic.com/research/reasoning-models-dont-say-think)

**Comment by SensenGao** [comment]
Thank you very much for your reply. 
I believe there is still a problem here. For example, when I was about to add 3 red balls, the result turned out to be incorrect—
it **matched the image, but it didn't follow the text**. Therefore, I think the inconsistency between the image and the text leads to errors in the answer.

<img width="1051" height="803" alt="Image" src="https://github.com/user-attachments/assets/daad1e06-a4e9-4ce8-917d-94ce6379d3b7" />

<img width="1387" height="806" alt="Image" src="https://github.com/user-attachments/assets/0995de29-c2c2-48da-9e1c-e551b090f862" />

**Comment by LeonLixyz** [comment]
Hi, thanks for providing more example. For your above issue, do you meant that i) the model generated 5 balls instead of 3 ii) the model's text output was + 5 instead of +3, which actually matched with what the image? 

This is an expected behavior, because the bagel model has 2 encoders: one from VAE and one from ViT. During interleaved generation, the image was feedback into the context with the ViT encoder as well, which caused the model to understand +5 balls (from the incorrect image), and lead to the incorrect text output. You can find more details about the architecture of Bagel [here](https://arxiv.org/pdf/2505.14683)

**Comment by SensenGao** [comment]
Thank you again for your answer. My point is that **if consistency in images cannot be ensured during multimodal reasoning, it often leads to incorrect results** and deviates from the original purpose of introducing visual reasoning into the reasoning process.

The fine-tuned BAGEL-Zebra-CoT often produces many errors in its reasoning images. I think this is an area that could be improved, possibly by increasing the amount of training data.
